man/dwfilter.1
.\" Generated by manscript from dwfilter.1.txt
.TH "DWFILTER" "1" "2013/03/10" "Version 2.0.9" "reformat text for processing"
.hw /usr/share/doc/dwdiff-2.0.9 http://os.ghalkes.nl/dwdiff.html
.SH NAME
dwfilter \- reformat text with dwdiff for further processing
.SH SYNOPSIS
\fBdwfilter\fR [\fIOPTIONS\fR] \fIOLD FILE\fR \fINEW FILE\fR \fIPOST PROCESSOR\fR [\fIPOST PROCESSOR OPTIONS\fR]
.SH DESCRIPTION
\fBdwfilter\fR reformats the text in the old file according to the contents of
the new file (or vice versa) and subsequently passes the reformated old file
and the new file through a secondary filter. It's main use is to allow visual
diff programs such as \fBmeld\fR and \fBkdiff3\fR to be used, eventhough a text
file has been reformated after editing. A further use is to allow the creation
of small patches even when the new text has been reformated. \fBdwfilter\fR
uses \fBdwdiff\fR for reformatting.
.SH OPTIONS
.TP
\fB\-r\fR, \fB\-\-reverse\fR
Reformat the new file based on the contents of the old file instead of the
default where the old file is reformated based on the contents of the new file.
.PP
\fBdwfilter\fR accepts the following \fBdwdiff\fR options:
.PP
.RS
\fB\-d\fR \fIdelimiters\fR, \fB\-\-delimiters\fR=\fIdelimiters\fR
.br
\fB\-P\fR, \fB\-\-punctuation\fR
.br
\fB\-W\fR \fIwhitespace\fR, \fB\-\-whitespace\fR=\fIwhitespace\fR
.br
\fB\-i\fR, \fB\-\-ignore\-case\fR
.br
\fB\-I\fR, \fB\-\-ignore\-formatting\fR
.br
\fB\-D\fR \fIoption\fR, \fB\-\-diff\-option\fR=\fIoption\fR
.br
\fB\-C\fR\fInum\fR, \fB\-\-context\fR=\fInum\fR
.br
\fB\-m\fR\fInum\fR, \fB\-\-match\-context\fR=\fInum\fR
.br
\fB\-\-aggregate\-changes\fR
.br
\fB\-\-wdiff\-output\fR
.RE
.PP
See the \fBdwdiff\fR manual page for the meaning.
.PP
A single dash (\-) as a file can be used to denote standard input. Only one
file can be read from standard input. To stop \fBdwfilter\fR from interpreting
file names that start with a dash as options, one can specify a double dash
(\-\-) after which \fBdwfilter\fR will interpret any following arguments as files
to read.
.SH BUGS
If you think you have found a bug, please check that you are using the latest
version of \fBdwdiff\fR <http://os.ghalkes.nl/dwdiff.html>. When reporting
bugs, please include a minimal example that demonstrates the problem.
.SH AUTHOR
G.P. Halkes <dwdiff@ghalkes.nl>
.SH COPYRIGHT
Copyright \(co 2006\-2010 G.P. Halkes
.br
dwdiff is licensed under the GNU General Public License version 3.
.br
For more details on the license, see the file COPYING in the documentation
directory. On Un*x systems this is usually /usr/share/doc/dwdiff-2.0.9.
.SH SEE\ ALSO
\fBdwdiff\fR(1), \fBdiff\fR(1), \fBmeld\fR(1), \fBkdiff3\fR(1)


Changelog
Version 2.0.9:
    Fix typo in configure script which prevents it from running.

Version 2.0.8:
    Bug fixes:
    - Fix build against glibc 2.16 and newer.

    Miscelaneous changes:
    - Added correct breaking for Unicode Regional Indicators.

Version 2.0.7:
    Fix incorrect release of 2.0.6, which omitted changes from the 2.0.5
    release.

Version 2.0.6
    Bug fixes:
    - Fix incorrect line numbering caused by optimized printing of differences
      preceeded in both files by whitespace containing newlines.

Version 2.0.5:
    Bug fixes:
    - fix a bug in the handling of context, which caused incorrect diffs to be
      printed. In some cases text was reported both as removed and as common.

Version 2.0.4:
    Bug fixes:
    - fix a memory allocation bug, which would be triggered by large values
      for the -m option.

Version 2.0.3:
    Updated translations for German (incomplete), French (incomplete) and
    Spanish.

Version 2.0.2:
    Bug fixes:
    - add support for DESTDIR in the Makefile install target.
    - fix build error on --without-unicode compiles.

Version 2.0.1:
    Bug fixes:
    - escape parsing for unknown escapes (like for example \;) would result
      in incorrect strings.
    - when the space character is not considered whitespace, dwdiff will no
      longer use it to make changes more readable.

    Internal changes:
    - the token file no longer uses a newline character as separator, but a nul
      character, making reparsing easier and the whole of dwdiff more robust.

Version 2.0:
    New features:
    - dwdiff is no longer a front-end for the diff program, but includes its
      own diff-algorithm implementation (taken from GNU diff). This makes
      dwdiff much faster, and ensures that results are the same across all
      platforms.

    Bug fixes:
    - when two additions are made with a single common token in between, dwdiff
      incorrectly reported the common token as replaced by the two additions
      and the common token.
    - when using the --diff-input option, in some cases dwdiff would report
      changes spanning multiple diff sections.

Version 1.9:
    New features:
    - an option for reading the output of diff (in Unified Diff) format was
      added (--diff-input) which can be used to reformat diffs for enhanced
      readability.

Version 1.8.2:
    New features:
    - when parsing options, \e is understood to mean the escape character

    Bug fixes:
    - the -p and -l options did not correctly handle UTF-8 characters.

Version 1.8.1:
    New features:
    - the context matching now decreases the matched context size exponentially,
      instead of immediately using the context-less mode if the changes found
      with context need refining. This results in more intuitive output when
      increasing the matching context size.

Version 1.8:
    New features:
    - added an option to repeat the begin and end markers at the start and end
      of line if a change crosses a newline

    Bug fixes:
    - newline characters in the delimiter set where handled incorrectly in
      UTF-8 mode

Version 1.7:
    New features:
    - added dwfilter program which reformats the old text such that it matches
      the new text to allow other programs such as visual diff tools to show
      the differences better. dwfilter can also be used to generate regular
      diffs to patch the old text by reformating the new text as the old text
      before executing diff [with caveats].

    Bug fixes:
    - \u and \U escapes were parsed incorrectly
    - dwdiff crashed when generating statistics for empty files

Version 1.6.1:
    Bug fixes:
    - the line clearing implemented for better background color handling caused
      (parts of) lines to disappear in the output with cr/nl line endings
    - the new paragraph separators were not handled correctly with the context
      matching feature, causing crashes

Version 1.6:
    New features:
    - the output for deleted text was changed such that the whitespace following
      the deletion is now the whitespace that followed the deleted text, rather
      than the whitespace that is now before the next word in the new file
    - the output for changed text was changed in the case that the whitespace
      before the old text contains a newline. In this case, depending on whether
      the whitespace before the new text contains a newline, the new file's
      whitespace is replaced by a single space or the new text is printed first
    - the --wdiff-output option was added to revert to the old output mode
    - the color specification now also allows specification of the background
      color and custom escape sequences
    - a new option to add markers for deleted or inserted blocks of lines
      containing only whitespace (paragraph separators)

    Bug fixes:
    - octal escapes are now properly parsed
    - icu-config was called directly rather than through $ICU_CONFIG in the
      configure script, which prevented proper operation of the --icu-config
      option
    - compilation on systems without nl_langinfo was fixed

Version 1.5.2
    New features:
    - better option parsing, which allows short options to be grouped. For
      example -iPc instead of -i -P -c

    Bug fixes:
    - compatibility fixes to configure script

Version 1.5.1:
    Bug fixes:
    - Statistics are now printed to stderr instead of stdout such that the only
      thing printed on stdout is the text
    - Repaired code for diff programs that do not support -a option
    - Fixed Makefile to pass settings for diff program to C compiler

Version 1.5:
    New features:
    - added match context which ensures better results for large edits with
      frequently occuring words

Version 1.4:
    New features:
    - an option to use all punctuation characters as delimiters
    - Unicode (UTF-8) support including formatting insensitive comparison

src/dwfilter.c
/* Copyright (C) 2010-2011 G.P. Halkes
   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License version 3, as
   published by the Free Software Foundation.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <errno.h>

#include "definitions.h"
#include "util.h"
#include "optionMacros.h"
#define DWFILTER_COMPILE
#include "optionDescriptions.h"

#define TEMPLATE "/tmp/dwdiffXXXXXXXX"
#define TEMPLATE_LENGTH sizeof(TEMPLATE)

#ifndef DWDIFF
#	define DWDIFF "dwdiff"
#endif


struct {
	int oldFile, newFile;
	bool reverse;
	int postProcessorStart;
} option;

static PARSE_FUNCTION(parseCmdLine)
	int noOptionCount = 0;
	bool discard;

	/* Initialise options to correct values */
	memset(&option, 0, sizeof(option));

	OPTIONS
		OPTION('d', "delimiters", REQUIRED_ARG)
		END_OPTION
		OPTION('P', "punctuation", NO_ARG)
		END_OPTION
		OPTION('W', "white-space", REQUIRED_ARG)
		END_OPTION
		OPTION('h', "help", NO_ARG)
				int i = 0;
				printf(_("Usage: dwfilter [OPTIONS] <OLD FILE> <NEW FILE> <POST PROCESSOR> [POST PROCESSOR OPTIONS]\n"));
				while (descriptions[i] != NULL)
					fputs(_(descriptions[i++]), stdout);
				exit(EXIT_SUCCESS);
		END_OPTION
		OPTION('v', "version", NO_ARG)
			fputs("dwfilter " VERSION_STRING "\n", stdout);
			fputs(
				/* TRANSLATORS:
				   - If the (C) symbol (that is the c in a circle) is not available,
					 leave as it as is. (Unicode code point 0x00A9)
				   - G.P. Halkes is name and should be left as is. */
				_("Copyright (C) 2006-2011 G.P. Halkes\nLicensed under the GNU General Public License version 3\n"), stdout);
			exit(EXIT_SUCCESS);
		END_OPTION
		OPTION('i', "ignore-case", NO_ARG)
		END_OPTION
		OPTION('I', "ignore-formatting", NO_ARG)
		END_OPTION
		SINGLE_DASH
		switch (noOptionCount++) {
			case 0:
				option.oldFile = optargind;
				break;
			case 1:
				option.newFile = optargind;
				break;
			default:
				option.postProcessorStart = optargind;
				STOP_OPTION_PROCESSING;
				break;
		}
		END_OPTION
		DOUBLE_DASH
			NO_MORE_OPTIONS;
		END_OPTION
		OPTION('C', "context", REQUIRED_ARG)
		END_OPTION
		OPTION('m', "match-context", REQUIRED_ARG)
		END_OPTION
		BOOLEAN_LONG_OPTION("aggregate-changes", discard)
		OPTION('A', "algorithm", REQUIRED_ARG)
		END_OPTION
		BOOLEAN_LONG_OPTION("wdiff-output", discard)
		/* FIXME: make this work again, after fixing dwdiff */
/* 		OPTION('S', "paragraph-separator", OPTIONAL_ARG)
		END_OPTION */

		BOOLEAN_OPTION('r', "reverse", option.reverse)

		fatal(_("Option %.*s does not exist\n"), OPTPRARG);
	NO_OPTION
		switch (noOptionCount++) {
			case 0:
				option.oldFile = optargind;
				break;
			case 1:
				option.newFile = optargind;
				break;
			default:
				option.postProcessorStart = optargind;
				STOP_OPTION_PROCESSING;
				break;
		}
	END_OPTIONS
	/* Check that we have something to work with. */
	if (noOptionCount < 2)
		fatal(_("Need two files to compare\n"));
	if (noOptionCount < 3)
		fatal(_("No post processor specified\n"));

END_FUNCTION

static char tempfile[TEMPLATE_LENGTH];

static void cleanup(void) {
	unlink(tempfile);
}

static void createTempfile(void) {
	int fd;
	/* Create temporary file. */
	/* Make sure the umask is set so that we don't introduce a security risk. */
	umask(~S_IRWXU);
	/* Make sure we will remove temporary files on exit. */
	atexit(cleanup);

	strcpy(tempfile, TEMPLATE);
	if ((fd = mkstemp(tempfile)) < 0)
		fatal(_("Could not create temporary file: %s\n"), strerror(errno));
	close(fd);
}

static int execute(char * const args[]) {
	pid_t pid;
	int status;

	if ((pid = fork()) == 0) {
		execvp(args[0], args);
		exit(127);
	} else if (pid < 0) {
		fatal(_("Could not execute %s: %s\n"), args[0], strerror(errno));
	}
	pid = wait(&status);
	if (pid < 0)
		fatal(_("Error waiting for child process to terminate: %s\n"), strerror(errno));
	return WEXITSTATUS(status);
}

#define OUTPUT_OPTION "--dwfilter="

int main(int argc, char *argv[]) {
	char **cmdArgs;
	char *outputArg;
	int i, j;

#if defined(USE_GETTEXT) || defined(USE_UNICODE)
	setlocale(LC_ALL, "");
#endif

#ifdef USE_GETTEXT
	bindtextdomain("dwdiff", LOCALEDIR);
	textdomain("dwdiff");
#endif

	parseCmdLine(argc, argv);
	createTempfile();
	cmdArgs = safe_malloc(sizeof(char *) * (option.postProcessorStart + 7));
	outputArg = safe_malloc(sizeof(char) * (strlen(tempfile) + sizeof(OUTPUT_OPTION)));

	cmdArgs[0] = safe_strdup(DWDIFF);
	/* argument has been allocated to correct size above, so we can safely use sprintf */
	sprintf(outputArg, OUTPUT_OPTION "%s", tempfile);
	cmdArgs[1] = outputArg;

	for (i = 1, j = 2; i < option.postProcessorStart; i++)
		if (i != option.oldFile && i != option.newFile)
			cmdArgs[j++] = argv[i];

	cmdArgs[j++] = safe_strdup("-w");
	cmdArgs[j++] = safe_strdup("");
	cmdArgs[j++] = safe_strdup("-x");
	cmdArgs[j++] = safe_strdup("");
	cmdArgs[j++] = safe_strdup("-2");

	cmdArgs[j++] = argv[option.reverse ? option.newFile : option.oldFile];
	cmdArgs[j++] = argv[option.reverse ? option.oldFile : option.newFile];
	cmdArgs[j++] = NULL;

	if (execute(cmdArgs) > 2)
		fatal(_("dwdiff returned an error\n"));

	free(outputArg);
	free(cmdArgs);

	cmdArgs = safe_malloc(sizeof(char *) * (argc - option.postProcessorStart + 3));

	for (i = option.postProcessorStart, j = 0; i < argc; i++, j++)
		cmdArgs[j] = argv[i];

	cmdArgs[j++] = option.reverse ? argv[option.oldFile] : tempfile;
	cmdArgs[j++] = option.reverse ? tempfile : argv[option.newFile];
	cmdArgs[j++] = NULL;

	return execute(cmdArgs);
}


